drivers/amazonec2/awscredentials.go

```diff
@@ -1,19 +1,79 @@
 package amazonec2
 
-import "github.com/aws/aws-sdk-go/aws/credentials"
+import (
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+)
 
 type awsCredentials interface {
-	NewStaticCredentials(id, secret, token string) *credentials.Credentials
+	Credentials() *credentials.Credentials
+}
+
+type ProviderFactory interface {
+	NewStaticProvider(id, secret, token string) credentials.Provider
+
+	NewSharedProvider(filename, profile string) credentials.Provider
+}
+
+type defaultAWSCredentials struct {
+	AccessKey        string
+	SecretKey        string
+	SessionToken     string
+	Filename         string
+	Profile          string
+	providerFactory  ProviderFactory
+	fallbackProvider awsCredentials
+}
+
+func NewAWSCredentials(id, secret, token, filename, profile string) *defaultAWSCredentials {
+	creds := defaultAWSCredentials{
+		AccessKey:        id,
+		SecretKey:        secret,
+		SessionToken:     token,
+		Filename:         filename,
+		Profile:          profile,
+		fallbackProvider: &AwsDefaultCredentialsProvider{},
+		providerFactory:  &defaultProviderFactory{},
+	}
+	return &creds
+}
+
+func (c *defaultAWSCredentials) Credentials() *credentials.Credentials {
+	providers := []credentials.Provider{}
+	if c.AccessKey != "" && c.SecretKey != "" {
+		providers = append(providers, c.providerFactory.NewStaticProvider(c.AccessKey, c.SecretKey, c.SessionToken))
+	}
+	if c.Filename != "" || c.Profile != "" {
+		providers = append(providers, c.providerFactory.NewSharedProvider(c.Filename, c.Profile))
```

Comment:
In the [SDK](https://github.com/aws/aws-sdk-go/blob/master/aws/defaults/defaults.go#L91), `EnvProvider` has precedence over `SharedCredentialsProvider`. If there is a reason for doing things differently here, I suggest it is properly explained in comments.